lib/backends/sdc/containers.js

```diff
@@ -1867,72 +2180,181 @@ function getContainerCount(opts, callback) {
+                cb();
+                return;
             }
-            return 0;
-        });
 
-        if (opts.since) {
+            objects = stash.objects;
+
             match = findContainerIdMatch(opts.since, objects);
             if (!match) {
-                callback(new errors.DockerError(
-                    'Could not find container with name or id '
-                    + opts.since));
+                cb(new errors.DockerError('Could not find container with name '
+                    + 'or id ' + opts.since));
+                return;
+            }
+
+            stash.objects = objects.slice(0, objects.indexOf(match));
+
+            cb();
+        }, function _filterBefore(stash, cb) {
+            var match;
+            var objects = stash.objects;
+
+            if (!opts.before) {
+                cb();
                 return;
             }
-            objects = objects.slice(0, objects.indexOf(match));
-        }
 
-        if (opts.before) {
             match = findContainerIdMatch(opts.before, objects);
             if (!match) {
-                callback(new errors.DockerError(
-                    'Could not find container with name or id '
-                    + opts.before));
+                cb(new errors.DockerError('Could not find container with name '
+                    + 'or id ' + opts.before));
+                return;
+            }
+            stash.objects = objects.slice(objects.indexOf(match) + 1);
+
+            cb();
+        }, function _preloadPackages(stash, cb) {
+            var filters;
+            var filteringOnPkgs = false;
+            var filterKeys;
+            var invalidPackageName;
+
+            if (opts.filters) {
+                filters = JSON.parse(opts.filters);
+                filterKeys = Object.keys(filters);
+            }
+
+            if (filters && filterKeys.indexOf('label') !== -1) {
+                filterKeys.forEach(function _findPkgSelectLabel(field) {
+                    var i;
+                    var labelSplit;
+                    var val = filters[field];
+
+                    for (i = 0; i < val.length; i++) {
+                        labelSplit = val[i].split('=', 2);
+                        if (labelSplit[0] === PACKAGE_SELECTION_LABEL) {
+                            filteringOnPkgs = true;
+                            if (labelSplit[1].match(BAD_PKG_NAME_RE)
+                                || !labelSplit[1].match(PKG_NAME_RE)) {
+                                // invalid package name, this is an error
+                                invalidPackageName = labelSplit[1];
+                            }
+                        }
+                    }
+                });
+            }
+
+            if (invalidPackageName) {
+                cb(new errors.DockerError('invalid value for '
+                    + PACKAGE_SELECTION_LABEL + ': ' + invalidPackageName));
+                return;
+            }
+
+            loadPackages({
```

Comment:
Originally it was not, but because of the fact that /containers/json includes the Labels for all packages, we need the map of uuid -> package name later on around line 2450. I realize that loading from PAPI each time adds some more complexity for now, but I've created DOCKER-687 and added a note about that above in loadPackages() so that we at least add caching of this data. The package data should be pretty static so loading it periodically even might be enough.